### PR TITLE
Less aggressive predicate flattening when not all sides are the same

### PR DESCRIPTION
**Pull request by vbekiaris**
Less aggressive predicate flattening when not all sides are the same
Adds SQL predicate tests

(supersedes #8792)

### CODE REVIEW

File: hazelcast/src/test/java/com/hazelcast/query/SqlPredicateTest.java [hunk 1]
```diff
@@ -67,13 +67,100 @@
...
+            };
+
+    @Test
+    public void testSqlPredicates() {
+        Employee employee = new Employee("Joe", "AUSTIN", 25, true, 0);
+        for (String s : TEST_MATCHING_SQL_PREDICATES) {
+            try {
+                assertSqlTrue(s, employee);
+            }
+            catch (AssertionError assertionError) {
```
Comment (jerrinot) [inline hazelcast/src/test/java/com/hazelcast/query/SqlPredicateTest.java]: I'd do following:
- rename `assertSqlTrue()` to `assertSqlMatching()`.
- amend `assertSqlMatching()` to include the SQL on failure: `assertTrue("Failed matching SQL predicate \"" + s + '"', predicate.apply(entry));`
- remote the try/catch block from here - no need to polute a test with it
- do the same for `assertSqlFalse()`
WDYT?

**Comment by jerrinot** [comment]
It looks good. I made a minor comment regarding a test purity. 

It can be merged and backported into a maintenance branch once it's green. Also please create a GitHub issue and set a milestone to 3.7.1 - so the fix will be visible in the release notes.

**Comment by jerrinot** [comment]
I sent backport #8798. The commits are just cherry-picked from this PR. 
However for master the tests should be cleaned a bit.

**Comment by vbekiaris** [comment]
Thanks for the test cleanup suggestion, looks much better now.

**Comment by jerrinot** [comment]
:+1:

File: hazelcast/src/main/java/com/hazelcast/query/SqlPredicate.java [hunk 2]
```diff
@@ -317,7 +317,7 @@ private void readObject(java.io.ObjectInputStream in)
         Predicate[] predicates;
         if (klass.isInstance(predicateLeft)) {
             subpredicatesLeft = ((CompoundPredicate) predicateLeft).getPredicates();
-            if (predicateRight instanceof CompoundPredicate) {
+            if (predicateRight.getClass() == klass) {
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/query/SqlPredicate.java]: I'm not familiar with the problem, so I don't understand the solution. Could you shine some light on it so I can give my thumbs up?

Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/query/SqlPredicate.java]: I intend to add a couple of tests before we get this merged, so I will mark as 'dont merge' this PR for now. The problem was that sub-predicates joined by different logical operator (AND / OR) would be eventually flattened to a single compound predicate, discarding one of the two operators, leading to incorrect representation of the SQL statement. The fixed condition ensures we only flatten when operator is the same.

Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/query/SqlPredicate.java]: This logic seemed a bit complicated to me so I did an attempt at refactoring, might be useful, might not...

```
        Predicate[] predicates;
        if (klass.isInstance(predicateLeft) || klass.isInstance(predicateRight)) {
            Predicate[] left = getSubPredicatesIfClass(predicateLeft, klass);
            Predicate[] right = getSubPredicatesIfClass(predicateRight, klass);
            
            predicates = new Predicate[left.length + right.length];
            ArrayUtils.concat(left, right, predicates);
        } else {
            predicates = new Predicate[] {predicateLeft, predicateRight};
        }
```

And the helper method : 

```
    private static <T extends CompoundPredicate> Predicate[] getSubPredicatesIfClass(Predicate predicate, Class<T> klass){
        if (klass.isInstance(predicate)) {
            return ((CompoundPredicate) predicate).getPredicates();
        } else {
            return new Predicate[]{predicate};
        }
    }
```

**Comment by vbekiaris** [comment]
Updated with refactored code by @mmedenjak and additional tests to verify flattening with combinations of `OR`/`AND` predicates

**Comment by vbekiaris** [comment]
run-lab-run

**Pull request merged**